lib/saluki-components/src/encoders/datadog/logs/mod.rs

```diff
@@ -0,0 +1,287 @@
+use async_trait::async_trait;
+use http::{uri::PathAndQuery, HeaderValue, Method, Uri};
+use memory_accounting::{MemoryBounds, MemoryBoundsBuilder};
+use saluki_common::iter::ReusableDeduplicator;
+use saluki_config::GenericConfiguration;
+use saluki_context::tags::Tag;
+use saluki_core::{
+    components::{encoders::*, ComponentContext},
+    data_model::{
+        event::{
+            log::{Log, LogStatus},
+            Event, EventType,
+        },
+        payload::{HttpPayload, Payload, PayloadMetadata, PayloadType},
+    },
+    observability::ComponentMetricsExt as _,
+    topology::PayloadsDispatcher,
+};
+use saluki_error::{ErrorContext as _, GenericError};
+use saluki_io::compression::CompressionScheme;
+use saluki_metrics::MetricsBuilder;
+use serde::Deserialize;
+use serde_json::{Map as JsonMap, Value as JsonValue};
+use tracing::{error, warn};
+
+use crate::common::datadog::{
+    io::RB_BUFFER_CHUNK_SIZE,
+    request_builder::{EndpointEncoder, RequestBuilder},
+    telemetry::ComponentTelemetry,
+    DEFAULT_INTAKE_COMPRESSED_SIZE_LIMIT, DEFAULT_INTAKE_UNCOMPRESSED_SIZE_LIMIT,
+};
+
+const DEFAULT_SERIALIZER_COMPRESSOR_KIND: &str = "zstd";
+const MAX_LOGS_PER_PAYLOAD: usize = 1000;
+
+static CONTENT_TYPE_JSON: HeaderValue = HeaderValue::from_static("application/json");
+
+fn default_serializer_compressor_kind() -> String {
+    DEFAULT_SERIALIZER_COMPRESSOR_KIND.to_owned()
+}
+
+const fn default_zstd_compressor_level() -> i32 {
+    3
+}
+
+fn default_ddsource() -> String {
+    "otel".to_string()
+}
+
+/// Datadog Logs incremental encoder.
+#[derive(Deserialize, Debug)]
+pub struct DatadogLogsConfiguration {
+    /// Compression kind for Logs payloads. Defaults to `zstd`.
+    #[serde(
+        rename = "serializer_compressor_kind",
+        default = "default_serializer_compressor_kind"
+    )]
+    compressor_kind: String,
+
+    /// Compressor level to use when the compressor kind is `zstd`. Defaults to 3.
+    #[serde(
+        rename = "serializer_zstd_compressor_level",
+        default = "default_zstd_compressor_level"
+    )]
+    zstd_compressor_level: i32,
+
+    /// Value for the `ddsource` field in Agent JSON envelope. Defaults to "otel".
+    #[serde(default = "default_ddsource")]
+    ddsource: String,
```

Comment:
Seems like we should have a generic `source` field on the `Log` type itself, since we should be able to (theoretically) run any sort of log through this encoder and have an appropriate source identifier for them.

As is, we would need one encoder per log source, which is a big limitation.